frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -28,6 +28,43 @@ static const std::vector<const IR::IDeclaration *> empty;
 
 ResolutionContext::ResolutionContext() { anyOrder = P4CContext::get().options().isv1(); }
 
+Util::Enumerator<const IR::IDeclaration *> *ResolutionContext::getDeclarations(
+    const IR::INamespace *ns) const {
+    auto nsIt = namespaceDecls.find(ns);
+    auto *declVec = nsIt != namespaceDecls.end() ? nsIt->second : memoizeDeclarations(ns);
+    return Util::Enumerator<const IR::IDeclaration *>::createEnumerator(*declVec);
+}
+
+std::vector<const IR::IDeclaration *> *ResolutionContext::memoizeDeclarations(
+    const IR::INamespace *ns) const {
+    auto decls = ns->getDeclarations();
+    if (auto nest = ns->to<IR::INestedNamespace>()) {
+        // boost bug -- trying to iterate with an adaptor over an unnamed temp crashes
+        auto temp = nest->getNestedNamespaces();
+        for (auto nn : boost::adaptors::reverse(temp)) decls = nn->getDeclarations()->concat(decls);
+    }
+
+    return (namespaceDecls[ns] = decls->toVector());
+}
+
+Util::Enumerator<const IR::IDeclaration *> *ResolutionContext::getDeclsByName(
+    const IR::INamespace *ns, cstring name) const {
+    auto nsIt = declNames.find(ns);
+    std::function<const IR::IDeclaration *(
+        const std::pair<const cstring, const IR::IDeclaration *> &)>
+        second = [](const auto &entry) { return entry.second; };
+
+    if (nsIt != declNames.end()) {
+        auto range = nsIt->second.equal_range(name);
+        return Util::enumerate(range.first, range.second)->map(second);
+    }
+
+    for (auto *d : *getDeclarations(ns)) declNames[ns].emplace(d->getName().name, d);
+
+    auto range = declNames[ns].equal_range(name);
+    return Util::enumerate(range.first, range.second)->map(second);
```

Comment:
There is `Values` in `lib/map.h` that should do a similar thing and might be better. You might need to create an enumerator from it since the rest of code uses enumerators.

Reply:
I'm not sure I understand, are you saying to try:
```
return Util::enumerate(Values(range).begin(), Values(range).end())->map(second);
```
?

Reply:
The point is that you would need `second` because that is exactly what the `Values` adaptor does. But since we need the enumerator anyway, this does not help much in the end.